### PR TITLE
Prefs[TASK[28626] Create UI (Detail Hotel Screen)

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 28626: Create UI (Detail Hotel Screen)](https://edu-redmine.sun-asterisk.vn/issues/28626)

## WHAT
- Create UI (Detail Hotel Screen)
## HOW
- Create UI (Detail Hotel Screen)
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/data/models/Hotel.kt [hunk 1]
```diff
@@ -18,7 +18,8 @@ data class Hotel(
     var price: String = "",
     var descriptionHotel: String = "",
     var rating_count: Double = 0.0,
-    var type: String = ""
+    var type: String = "",
+    var listImage: MutableList<String> = mutableListOf()
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/data/models/Hotel.kt]: Khi nào thì cần dùng `MutableList` thế? Và ô muốn thay cả list à mà có `var`

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/DetailImageAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun instantiateItem(parent: ViewGroup, position: Int): View =
+        LayoutInflater.from(parent.context)
+            .inflate(R.layout.item_image_slider, parent, false)
+            .also {
+                parent.addView(it)
+                it.imageViewSliderItem.imageViewSliderItem.loadImage(imageList[position])
+            }
+
+    override fun isViewFromObject(view: View, ob: Any) = view === ob as View
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/adapter/DetailImageAdapter.kt]: Có cần `cast` sang `View` khi so sánh 2 địa chỉ ô nhớ k?

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelAdapter.kt [hunk 3]
```diff
@@ -6,9 +6,9 @@ import com.sunasterisk.travelapp.R
 import com.sunasterisk.travelapp.data.models.Hotel
 import com.sunasterisk.travelapp.ui.base.BaseRecyclerAdapter
 
-class HotelAdapter : BaseRecyclerAdapter<Hotel, HotelViewHolder>() {
-
-    var onItemClick: (Hotel) -> Unit = { _, -> }
+class HotelAdapter(
+    private var onItemClick: (Hotel) -> Unit
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelAdapter.kt]: `private var` có nghĩa là cái adapter nó có thể thay đổi `onItemClick` à?

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelDestinationsAdapter.kt [hunk 4]
```diff
@@ -7,7 +7,7 @@ import com.sunasterisk.travelapp.data.models.Location
 import com.sunasterisk.travelapp.ui.base.BaseRecyclerAdapter
 
 class HotelDestinationsAdapter(
-    var onItemClick: (Location) -> Unit
+    private var onItemClick: (Location) -> Unit
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelDestinationsAdapter.kt]: Chuyển qua `val`

File: app/src/main/java/com/sunasterisk/travelapp/ui/detail/HotelDetailActivity.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.sunasterisk.travelapp.data.models.Hotel
+import com.sunasterisk.travelapp.ui.adapter.DetailImageAdapter
+import com.sunasterisk.travelapp.ui.base.BaseActivity
+import com.sunasterisk.travelapp.ui.list.hotel.HotelListActivity.Companion.EXTRA_HOTEL_ITEM
+import kotlinx.android.synthetic.main.activity_hotel_detail.*
+
+class HotelDetailActivity : BaseActivity() {
+
+    override val layoutResource get() = R.layout.activity_hotel_detail
+    private val adapter: DetailImageAdapter = DetailImageAdapter()
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/detail/HotelDetailActivity.kt]:
```suggestion
    private val adapter = DetailImageAdapter()
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt [hunk 6]
```diff
@@ -57,4 +61,19 @@ class HotelListActivity : BaseActivity() {
         if (item.itemId == android.R.id.home) finish()
         return super.onOptionsItemSelected(item)
     }
+
+    private fun onClick(item: Hotel) {
+        this.apply {
+            val intent = getDetailIntent(this, item)
+            startActivity(intent)
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt]:
```suggestion
        getDetailIntent(this, item).let(this::startActivity)
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt [hunk 7]
```diff
@@ -57,4 +61,19 @@ class HotelListActivity : BaseActivity() {
...
+    }
+
+    companion object {
+        const val EXTRA_HOTEL_ITEM = "EXTRA_HOTEL_ITEM"
+        fun getDetailIntent(context: Context, hotel: Hotel) =
+            Intent(context, HotelDetailActivity::class.java).apply {
+                putExtra(EXTRA_HOTEL_ITEM, hotel)
+            }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt]: Chuyển đến `companion` của `HotelDetailActivity`

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -32,4 +33,11 @@
...
+    \n\n\t\tPhasellus in aliquet mi. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. In volutpat arcu ut felis sagittis, in finibus massa gravida. Pellentesque id tellus orci. Integer dictum, lorem sed efficitur ullamcorper, libero justo consectetur ipsum, in mollis nisl ex sed nisl. Donec maximus ullamcorper sodales. Praesent bibendum rhoncus tellus nec feugiat. In a ornare nulla. Donec rhoncus libero vel nunc consequat, quis tincidunt nisl eleifend. Cras bibendum enim a justo luctus vestibulum. Fusce dictum libero quis erat maximus, vitae volutpat diam dignissim.
+  </string>
```
Comment (aug16vcc) [inline app/src/main/res/values/strings.xml]: Dùng `<TextView ... tools:text="@tools:sample/..." />`

**Review by aug16vcc — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/data/models/Hotel.kt (hunk 1 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/data/models/Hotel.kt]: MutableList extends List, có thêm một số hàm thêm add/remove/update -> dùng khi muốn thay đổi thay đổi các giá trị của nó

Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/data/models/Hotel.kt]: ok, ông dùng MutableList rồi thì ko cần dùng var nữa

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelAdapter.kt (hunk 3 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/adapter/HotelAdapter.kt]: ok ông, t chuyển qua val

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/DetailImageAdapter.kt (hunk 2 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/adapter/DetailImageAdapter.kt]: Ok ông nếu ob ko phải View, sẽ bị crash do throw CatsException thay vì trả về false (hiển thị blank view)

File: app/src/main/java/com/sunasterisk/travelapp/ui/detail/HotelDetailActivity.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    private fun setUpHotelImageSlider(imageList: MutableList<String>) {
+        viewPagerHotelThumb.adapter = adapter
+        adapter.updateList(imageList)
+    }
+
+    companion object {
+        const val EXTRA_HOTEL_ITEM = "EXTRA_HOTEL_ITEM"
+        fun getDetailIntent(context: Context, hotel: Hotel) =
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/detail/HotelDetailActivity.kt]:
```suggestion
        fun getIntent(context: Context, hotel: Hotel) =
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt [hunk 10]
```diff
@@ -57,4 +59,8 @@ class HotelListActivity : BaseActivity() {
         if (item.itemId == android.R.id.home) finish()
         return super.onOptionsItemSelected(item)
     }
+
+    private fun onClick(item: Hotel) {
+        getDetailIntent(this, item).let(this::startActivity)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/list/hotel/HotelListActivity.kt]:
```suggestion
        HotelDetailActivity.getIntent(this, item).let(this::startActivity)
```

**Review by aug16vcc — CHANGES_REQUESTED** [review]

**Review by aug16vcc — APPROVED** [review]

File: app/src/main/res/layout/activity_hotel_detail.xml [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                android:text="@string/title_detail"
+                android:textAppearance="@style/TextAppearance.MaterialComponents.Subtitle1"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toBottomOf="@+id/textViewPrice" />
+
+            <TextView
+                android:layout_width="0dp"
+                android:layout_height="wrap_content"
+                android:layout_margin="@dimen/dp_16"
+                android:layout_marginTop="12dp"
```
Comment (NguyenTaTrung) [inline app/src/main/res/layout/activity_hotel_detail.xml]: hard code

File: app/src/main/res/layout/activity_hotel_detail.xml [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+    <com.google.android.material.floatingactionbutton.FloatingActionButton
+        style="@style/Widget.TravelApp.Button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="end|bottom"
+        android:layout_margin="@dimen/dp_16"
+        android:src="@drawable/ic_book"
+        app:tint="@color/color_primary" />
+</androidx.coordinatorlayout.widget.CoordinatorLayout>
```
Comment (NguyenTaTrung) [inline app/src/main/res/layout/activity_hotel_detail.xml:130]: format code

Comment (NguyenTaTrung) [reply app/src/main/res/layout/activity_hotel_detail.xml:130]: Lỗi nhỏ thôi, ông có thể sửa luôn hoặc pull sau nhé

Comment (hieuna-441907) [reply app/src/main/res/layout/activity_hotel_detail.xml:130]: ok ông, t sửa luôn

**Review by thonc-2135 — APPROVED** [review]

**Pull request merged**